### ISSUE TITLE
Document rough memory/disk usage requirements

### ISSUE THREAD

**Issue by tsibley** [etsy/Hound#15]
Hound is super-simple to get running, which is awesome!  I'm trying to index a number of repos on a host that's got lots of disk, but little memory. (1GB... I know. I want to increase that but can't at the moment for Reasons.)  The indexer runs for a brief period and then hits a panic:

   panic: fork/exec /usr/bin/git: cannot allocate memory

My currently configured repos (100 of em) are just a small fraction of the total I'd like to index.  I'll be moving to a different host, but I'm wondering if you have any rough rules of thumb for resource requirements.  Obviously this is very repo dependent, but perhaps something normalized like a ratio of source repo size on disk to index size and indexer memory requirements?

**Comment by jklein** [comment]
Yeah this is a good idea. We could probably figure out a rough range based on repo size and put it in the README. Or even put more details in the output.

**Comment by tsibley** [comment]
At least on the small sample of 100 repos I tried, the disk usage seems to be about a 1:2 ratio between `data/<repo>-<sha>/raw` and `data/<repo>-<sha>/tri`.

**Comment by kellegous** [comment]
I'll have to think more about a good way to quantify the memory/disk requirements. The most memory intensive part of Hound is indexing. This is the only time when the index structure is resident in the program's heap. The index consists of the posting list for each trigram and the list of file paths. The index itself is usually a lot smaller than the original repository ... like 5x-10x smaller. Once that structure is written to disk, though, access happens via mmap. This should mean that heap usage doesn't really grow linearly with the number of repositories you add. Obviously, having memory helps as you really want mmap operating out of the page cache but in the very worst case, mmap is just taking address space in the process and not actual heap. I'm quite surprised to see that git is what ran out of memory for you. Are you able to clone the repository that failed on this machine even when Hound isn't running?

**Comment by tsibley** [comment]
Yes, all of the repos can be cloned by git.  It isn't git itself running out of memory, but that the exec() syscall Hound makes fails because the system says there's not enough memory left (i.e. because Hound was claiming it).

**Comment by tsibley** [comment]
The exec() failure happens at different points, but if I keep houndd in a retry loop and bump up the repo update delay (to avoid unnecessary contention), the initial indexing eventually succeeds and houndd continues to run fine with limited memory.  I assume the panic arises when the initial clone/pull/index for each repo stack up enough at the same time that the in-heap indexes max out all available memory.

Anyway, memory-constrained environments aren't necessarily worth optimizing for.  I just thought it'd be useful to document the expected rough usages, as it wasn't clear from the beginning that houndd would keep functioning even if I could get initial indexing to succeed.

**Comment by MoriTanosuke** [comment]
I'm getting the same error when checking out about 75 SVN repositories:

    2015/02/05 14:40:24 vcs pull error (REPONAME - https://REPOURL/trunk): fork /exec /usr/bin/svn: cannot allocate memory

The machine I'm using has 2GB RAM, where about 300MB are still available and has 50% available harddive space.

@tsibley Can you describe how to modify the repo update delay?

**Comment by MoriTanosuke** [comment]
To answer my own question, adding `'MsBetweenPolls': 180000` to each repo configuration sets the poll intervall to 180s.

**Comment by tsibley** [comment]
@MoriTanosuke I think the key is `ms-between-poll`, not `MsBetweenPolls`.  See https://github.com/etsy/Hound/blob/master/config-example.json#L9.
